blockstack/stacks-blockchain
# Use RBF for recovering from fast blocks

This PR is superseding  https://github.com/blockstack/stacks-blockchain/pull/2363 and addressing a few issues, with one refactoring;
- It allows a miner to recover from a "fast block", by RBF-ing submitted block commits trying to target a "flashed" block.
- It's preventing submission of multiple block commit / block. 
- Improving the RBF logic, by dynamically computing the size of the transactions (the values from config are still used but now only for initial estimation).

I was able to test the miner against a puppet chain and see the RBF logic successfully kicking in with multiple consecutive flash blocks (3 consecutive RBF, 1 final mined transaction), I should also be able to write some unit tests.


How to test:
1) start bitcoind
```bash
$ cat helium.conf
chain=regtest
disablewallet=0
txindex=1
server=1
rpcuser=helium-node
rpcpassword=secret%

$ bitcoind -conf=helium.conf
```

2) start puppet chain
```bash
$ cd testnet/puppet-chain
$ DYNAMIC_GENESIS_TIMESTAMP=1 cargo run local-leader.toml.default
```

3) start a stacks-node

in `core/mod.rs` update the constants 
```rust
pub const BITCOIN_REGTEST_FIRST_BLOCK_HEIGHT: u64 = 300;
pub const BITCOIN_REGTEST_FIRST_BLOCK_TIMESTAMP: u32 = 1611158538;
pub const BITCOIN_REGTEST_FIRST_BLOCK_HASH: &str =
    "47922cbadd8158ea999b6be9a31dd69daf58b33dd5be2d3d2d528274d5dd8ea2";
```

I'd also recommend enforce usage of a sampled genesis.

```bash
$ cd testnet/stacks-node
$ STACKS_LOG_PP=1 cargo run start --config=./conf/local-leader-conf.toml
```

The puppet chain will basically be messing around with bitcoind between block 307 and block 312, generating flash blocks and buffering transactions coming from the stacks-node.
When everything goes back to normal and inspecting block 313, we can see that all the block commit that happened during 307 and 312 have been RBF, and we only have one clean block commit.
